pkg/collector/upgrade/upgrade.go

```diff
@@ -59,20 +61,36 @@ func ManagedInstances(ctx context.Context, logger logr.Logger, cl client.Client)
 }
 
 // ManagedInstance performs the necessary changes to bring the given otelcol instance to the current version
-func ManagedInstance(ctx context.Context, logger logr.Logger, cl client.Client, otelcol *v1alpha1.OpenTelemetryCollector) (*v1alpha1.OpenTelemetryCollector, error) {
-	if v, ok := versions[otelcol.Status.Version]; ok {
-		// we don't need to run the upgrade function for the version 'v', only the next ones
-		for n := v.next; n != nil; n = n.next {
-			// performs the upgrade to version 'n'
-			upgraded, err := n.upgrade(cl, otelcol)
+func ManagedInstance(ctx context.Context, logger logr.Logger, currentV version.Version, cl client.Client, otelcol v1alpha1.OpenTelemetryCollector) (v1alpha1.OpenTelemetryCollector, error) {
+	// this is likely a new instance, assume it's already up to date
+	if otelcol.Status.Version == "" {
+		otelcol.Status.Version = currentV.OpenTelemetryCollector
+		return otelcol, nil
+	}
+
+	instanceV, err := semver.NewVersion(otelcol.Status.Version)
+	if err != nil {
+		logger.Error(err, "failed to parse version for OpenTelemetry Collector instance", "name", otelcol.Name, "namespace", otelcol.Namespace, "version", otelcol.Status.Version)
+		return otelcol, err
+	}
+
+	if instanceV.GreaterThan(&Latest.Version) {
+		logger.Info("skipping upgrade for OpenTelemetry Collector instance, as it's newer than our latest version", "name", otelcol.Name, "namespace", otelcol.Namespace, "version", otelcol.Status.Version)
+		return otelcol, nil
+	}
+
+	for _, available := range versions {
+		if available.GreaterThan(instanceV) {
+			upgraded, err := available.upgrade(cl, &otelcol)
+
 			if err != nil {
 				logger.Error(err, "failed to upgrade managed otelcol instances", "name", otelcol.Name, "namespace", otelcol.Namespace)
 				return otelcol, err
 			}
 
-			logger.V(1).Info("step upgrade", "name", otelcol.Name, "namespace", otelcol.Namespace, "version", n.Tag)
-			upgraded.Status.Version = n.Tag
-			otelcol = upgraded
+			logger.V(1).Info("step upgrade", "name", otelcol.Name, "namespace", otelcol.Namespace, "version", available.String())
+			upgraded.Status.Version = available.String()
+			otelcol = *upgraded
 		}
 	}
```

Comment:
As the loop only includes versions that have an associated upgrade function, we still need to set the `upgraded.Status.Version` to `currentV.OpenTelemetryCollector` before exiting?